stream-video-android-core/src/main/kotlin/io/getstream/video/android/core/call/utils/TrackOverridesHandler.kt

```diff
@@ -0,0 +1,183 @@
+    private val onOverridesUpdate: (overrides: Map<String, TrackOverride>) -> Unit,
+    private val logger: TaggedLogger? = null,
+) {
+
+    private val trackOverrides: MutableMap<String, TrackOverride> = mutableMapOf()
+
+    internal data class TrackOverride(
+        val dimensions: VideoDimension? = null,
+        val visible: Boolean? = null,
+    )
+
+    /**
+     * Updates incoming video dimensions overrides.
+     *
+     * @param sessionIds List of session IDs to update. If `null`, the override will be applied to all participants.
+     * @param dimensions Video dimensions to set. Set to `null` to switch back to auto.
+     */
+    internal fun updateOverrides(
+        sessionIds: List<String>? = null,
+        dimensions: VideoDimension? = null,
+    ) {
+        val putDimensions = dimensions
+
+        if (sessionIds == null) {
+            putOrRemoveDimensionsOverride(
+                sessionId = ALL_PARTICIPANTS,
+                putDimensions = putDimensions,
+                existingVisibility = trackOverrides[ALL_PARTICIPANTS]?.visible,
+            )
+        } else {
+            sessionIds.forEach { sessionId ->
+                putOrRemoveDimensionsOverride(
+                    sessionId = sessionId,
+                    putDimensions = putDimensions,
+                    existingVisibility = trackOverrides[sessionId]?.visible,
+                )
+            }
+        }
+
+        onOverridesUpdate(trackOverrides)
+
+        logger?.d { "[updateOverrides] #manual-quality-selection; Overrides: $trackOverrides" }
+    }
+
+    private fun putOrRemoveDimensionsOverride(sessionId: String, putDimensions: VideoDimension?, existingVisibility: Boolean?) {
+        if (putDimensions == null && (existingVisibility == true || existingVisibility == null)) {
+            trackOverrides.remove(sessionId)
+        } else {
+            trackOverrides.put(
+                sessionId,
+                TrackOverride(dimensions = putDimensions, visible = existingVisibility),
+            )
+        }
+    }
+
+    /**
+     * Updates incoming video visibility overrides.
+     *
+     * @param sessionIds List of session IDs to update. If `null`, the override will be applied to all participants.
+     * @param visible Video visibility to set. Set to `null` to switch back to auto.
+     */
+    internal fun updateOverrides(
+        sessionIds: List<String>? = null,
+        visible: Boolean? = null,
+    ) {
+        val putVisibility = visible
+
+        if (sessionIds == null) {
+            putOrRemoveVisibilityOverride(
+                sessionId = ALL_PARTICIPANTS,
+                putVisibility = putVisibility,
+                existingDimensions = trackOverrides[ALL_PARTICIPANTS]?.dimensions,
+            )
+
+            getParticipantList().forEach {
+                val override = applyOverrides(it.sessionId, true)
+                it._videoEnabled.value = override
```

Comment:
We should not update the participant state here.
See https://github.com/GetStream/stream-video-android/pull/1242/files#r1869802104

Reply:
This was one of the points that I raised as a red flag and, after discussing it, I'll check if the video is enabled directly in the UI.